### PR TITLE
Subtitles for ggplot2

### PR DESCRIPTION
**Pull request by hrbrmstr**
- new `subtitle` parameter for `labs` and `ggtitle`
- documented the `subtitle` parameter in `ggtitle` since it's an explicit parameter
- new `caption` parameter for `labs` function
- new `plot.subtitle` theme element
- new `plot.caption` theme element (default alignment is right-justified)
- updated default alignment for `plot.title` and `plot.subtitle` (now left-justified by default)
- updated roxygen documentation with `subtitle` and `caption` examples
- added test for `title`, `subtitle` & `caption` settings
- cleaned up `=` spacing in a few other files since they came up when I was trying to fix my own :-)
- updated `NEWS.md` (including suggested fixes to `NEWS.md`)

Sample code to validate non-breakage (i.e. test with various plot elements):

```
library(ggplot2)
library(gridExtra)

st <- "Four score and seven years ago our fathers brought forth on this continent, a new nation, conceived in Liberty, and dedicated to the proposition that all people are created equal."
st <- paste0(strwrap(st, 70), sep="", collapse="\n")

p <- ggplot(mtcars, aes(wt, mpg))
p <- p + geom_point(aes(color=factor(cyl)))

# Changing the font solely to validate functionality. Times. Ugh.
p <- p + theme(text=element_text(family="Times"))
p <- p + theme(plot.title=element_text(face="bold"))
p <- p + theme(plot.subtitle=element_text(margin=margin(b=15)))
p <- p + theme(plot.caption=element_text(margin=margin(t=15),
                                        face="italic", size=8))
grid.arrange(
  
  p + ggtitle("Just ggtitle"),
  
  p + labs(title="Just labs with title"),
  
  p + ggtitle("Now ggtitle with title subtitle", subtitle=st),
  
  # More tests --------------------------------------------------------------
  
  p + labs(title="Now labs with title subtitle", subtitle=st) +
    theme(legend.position="top"),
  
  p + labs(title="Two separate labs calls for each title") +
    labs(subtitle=st) +
    theme(legend.position="bottom"),
  
  p + facet_wrap(~cyl) +
    labs(title="Works with facets too",
         subtitle=st,
         caption="Brought you by the letters 'gg'"),
  
  # below -------------------------------------------------------------------
  
  p + labs(title="Now labs with title subtitle & below",
           subtitle=st,
           source="Brought you by the letters 'gg'"),
  
  p + labs(title="Two separate labs calls for each title & below") +
    labs(subtitle=st) +
    labs(caption="Brought you by the letters 'gg'") +
    theme(legend.position="bottom"),
  
  p + facet_wrap(~cyl) +
    labs(title="Works with facets too & below",
         subtitle=st,
         caption="Brought you by the letters 'gg'"),
  
  ncol=3
  
)
```

image

### CODE REVIEW

File: R/labels.r [hunk 1]
```diff
@@ -58,8 +62,12 @@ ylab <- function(label) {
 }
 #' @rdname labs
 #' @export
-ggtitle <- function(label) {
-  labs(title = label)
+ggtitle <- function(label, subtitle) {
+  if (!missing(subtitle)) {
+    labs(title = label, subtitle = subtitle)
```
Comment (hadley) [inline R/labels.r]: Simpler to give `subtitle` a default value of `NULL`?

**Comment by hadley** [comment]
This is appealingly simple!

**Comment by JanSchulz** [comment]
[Follow up form twitter: https://twitter.com/hrbrmstr/status/709102677351911424]

@hrbrmstr The cowplot code is this: https://github.com/wilkelab/cowplot/blob/master/R/add_sub.R -> very similar to what you have here.

What I'm not sure is what to name it: adding it as an additional option to ggtitle does not feel right.

From the documentation in cowplot:
> #' Add annotation underneath a plot
> #' This function can add an arbitrary label or mathematical expression underneath
> #' the plot, similar to the \code{sub} parameter in base R.

File: R/theme.r [hunk 2]
```diff
@@ -254,6 +258,11 @@ print.theme <- function(x, ...) utils::str(x)
 #' p + theme(plot.title = element_text(size = rel(2)))
 #' p + theme(plot.title = element_text(size = rel(2), colour = "blue"))
 #'
+#' # Add a subtitle and adjust bottom margin
+#' p + labs(title = "Vehicle Weight-Gas Mileage Relationship",
+#'          subtitle = "You need to wrap long subtitleson manually") +
```
Comment (JanSchulz) [inline R/theme.r]: `s/subtitleson/subtitles/`?

**Comment by JanSchulz** [comment]
I added a PR for whole plot annotations: https://github.com/hrbrmstr/ggplot2/pull/1

File: R/theme.r [hunk 3]
```diff
@@ -187,7 +187,11 @@ print.theme <- function(x, ...) utils::str(x)
 #'   plot.background  \tab background of the entire plot
 #'                    (\code{element_rect}; inherits from \code{rect}) \cr
 #'   plot.title       \tab plot title (text appearance)
-#'                    (\code{element_text}; inherits from \code{title}) \cr
+#'                    (\code{element_text}; inherits from \code{title})
+#'                    left-aligned by default\cr
+#'   plot.subtitle    \tab plot title (text appearance)
```
Comment (JanSchulz) [inline R/theme.r]: s/plot title/plot subtitle/

**Comment by JanSchulz** [comment]
```
library(ggplot2)
library(ggplot2)
library(gridExtra)

st <- "Four score and seven years ago our fathers brought forth on this continent, a new nation, conceived in Liberty, and dedicated to the proposition that all men are created equal."
st <- paste0(strwrap(st, 70), sep="", collapse="\n")

p <- ggplot(mtcars, aes(wt, mpg))
p <- p + geom_point()

# Changing the font solely to validate functionality. Times. Ugh.
p <- p + theme(text=element_text(family="Times"))
p <- p + theme(plot.title=element_text(face="bold"))
p <- p + theme(plot.subtitle=element_text(margin=margin(b=15)))
p <- p + labs(title="Now labs with title subtitle", subtitle=st)

grid.arrange(
  p,
  p + labs(annotation="(based on mtcars dataset available in R)"),
  p + labs(annotation="(based on mtcars dataset available in R)") + 
    theme(plot.annotation=element_text(face="bold", hjust = 1)),
  ncol=3
)
```

rplot01

**Comment by hrbrmstr** [comment]
Incorporated the PR (tweaked the docs a bit). Fixed the Travis error. Rejiggered the visual validation (see new image above).

File: R/theme.r [hunk 4]
```diff
@@ -187,7 +187,14 @@ print.theme <- function(x, ...) utils::str(x)
 #'   plot.background  \tab background of the entire plot
 #'                    (\code{element_rect}; inherits from \code{rect}) \cr
 #'   plot.title       \tab plot title (text appearance)
-#'                    (\code{element_text}; inherits from \code{title}) \cr
+#'                    (\code{element_text}; inherits from \code{title})
+#'                    left-aligned by default\cr
+#'   plot.subtitle    \tab plot subtitle (text appearance)
+#'                    (\code{element_text}; inherits from \code{title})
+#'                    left-aligned by default\cr
+#'   plot.sub         \tab annotations below the plot (text appearance)
```
Comment (JanSchulz) [inline R/theme.r]: s/plot.sub/plot.annotation/...

Want a PR against your repo or simple do the change in another commit?

File: NEWS.md [hunk 5]
```diff
@@ -4,6 +4,15 @@
   aesthetics they accept: `xmin_final`, `xmax_final`, `xlower`,
   `xmiddle` and `xupper` are now valid `x` aesthetics.
 
+* `ggtitle` and `labs` now take a `subtitle` parameter which makes it 
+  possible to add a subtitle below the main plot title.
+  
+* `labs` now takes an `annotation` parameter that will set the label for
```
Comment (hadley) [inline NEWS.md]: "annotation" sounds way too generic for me. What about "source"?

File: NEWS.md [hunk 6]
```diff
@@ -4,6 +4,15 @@
...
   `xmiddle` and `xupper` are now valid `x` aesthetics.
 
+* `ggtitle` and `labs` now take a `subtitle` parameter which makes it 
+  possible to add a subtitle below the main plot title.
+  
+* `labs` now takes an `annotation` parameter that will set the label for 
+  annotation text below the plot panels
+  
+* the main plot title and subtitle are left-justified (`hjust = 0`) by
+  default, now
```
Comment (hadley) [inline NEWS.md]: You can drop ", now".

Can you please make it a proper sentence with leading capital and trailing period?

File: NEWS.md (hunk 5 above)
Comment (hadley) [inline NEWS.md]: Also my style is to add `()` to function names

File: R/labels.r [hunk 7]
```diff
@@ -51,15 +63,17 @@ labs <- function(...) {
...
 #' @export
 ylab <- function(label) {
   labs(y = label)
 }
+
 #' @rdname labs
 #' @export
-ggtitle <- function(label) {
-  labs(title = label)
+ggtitle <- function(label, subtitle=NULL) {
```
Comment (hadley) [inline R/labels.r]: Missing spaces around `=`

**Comment by hadley** [comment]
Looking good, just a few minor niggles left.

File: R/labels.r (hunk 7 above)
Comment (hrbrmstr) [inline R/labels.r]: will fix (remind me never to poke CRAN about their barriers to entry ;-)

File: R/theme-elements.r [hunk 8]
```diff
@@ -278,6 +278,8 @@ el_def <- function(class = NULL, inherit = NULL, description = NULL) {
 
   plot.background     = el_def("element_rect", "rect"),
   plot.title          = el_def("element_text", "title"),
+  plot.subtitle       = el_def("element_text", "title"),
+  plot.annotation     = el_def("element_text", "title"),
```
Comment (hadley) [inline R/theme-elements.r]: Should this default to right alignment?

Comment (hrbrmstr) [inline R/theme-elements.r]: that's a good q. if we set it to that by default and let folks play with
`devtools` version of ggplot2, we'll def get community feedback. I'll mod
that too.

**Comment by JanSchulz** [comment]
re "annotation" vs "source": They will probably be used for at least three use cases: source statements, copyright statements, and notices for explanations regarding the plots (depending on your style, the last could also go to the subtitle). That's why I settled on "annotations" which in my (non native) understanding encompass all three use cases.

**Comment by hadley** [comment]
Annotation is already used elsewhere in ggplot2 for a different purpose, so that name is definitely out. How about footer?

**Comment by hrbrmstr** [comment]
Lynn (Cherny) just suggested "caption" (I'm crowdsourcing this on Twitter
with some vis ppl :-)

**Comment by hadley** [comment]
Oh, I like "caption"!

**Comment by hrbrmstr** [comment]
I'll shall make said change when I'm off the train to Boston.

**Comment by JanSchulz** [comment]
Damn, wasn't fast enough: https://github.com/hrbrmstr/ggplot2/pull/3

File: NEWS.md [hunk 9]
```diff
@@ -4,6 +4,19 @@
   aesthetics they accept: `xmin_final`, `xmax_final`, `xlower`,
   `xmiddle` and `xupper` are now valid `x` aesthetics.
 
+* `ggtitle()` and `labs()` take a `subtitle` parameter which makes it
```
Comment (hadley) [inline NEWS.md]: One last thing - can you please credit yourself here, i.e. `(@hrbmstr)`

**Comment by JanSchulz** [comment]
This actually closes an issue which I opened in 2012: https://github.com/hadley/ggplot2/issues/682 :-)

File: NEWS.md (hunk 9 above)
Comment (hrbrmstr) [inline NEWS.md]: done.

**Comment by hadley** [comment]
Thanks!

**Comment by patternproject** [comment]
If I need to use this feature, should I pull the latest version from github or CRAN.

**Comment by ptoche** [comment]
Is this feature available? Thanks.

    devtools::install_github("tidyverse/ggplot2") 
    Error in ggsubtitle("TEST") : 
      could not find function "ggsubtitle"

**Comment by hrbrmstr** [comment]
You set it with `labs()`. There's plenty of help there, but `labs()` can
now take `subtitle=""` and `caption=""` params.

**Comment by ptoche** [comment]
Thanks! 

I liked the ``ggtitle()`` syntax for keeping things on single lines. Can I write something like this? (I'll try tomorrow, as I must be off to bed now)

    ggsubtitle <- function(x) {
      return(labs(subtitle = x))
    }
